src/types/paths.ts

```diff
@@ -0,0 +1,88 @@
+ *
+ *  - the first T extends T seems to be excessive, but the [keyof T] indexed access
+ *    types operate differently on unions - they only access the keys common to all the
+ *    members in the union. Some of our schema types are unions, so to capture all
+ *    possible access paths, we need to force distribution over each member in the union,
+ *    which is done through a conditional type
+ *
+ *  - once inside of that, we know that T is an object (or that's the expectation at
+ *    least) and we apply a mapped type - the keys are all possible keys in T, and each
+ *    optional key is made required (that's the `-?` modifier).
+ *
+ *  - the values in the mapped type are checked first with a conditional type - we only
+ *    care about string keys (because JSON serializable data is expected), so this does
+ *    not return array indices either (by design - we handle an array of values as a
+ *    single value in Formik), that's the `K extends string`.
+ *
+ *    When we have a string key, we check if we need to recurse by looking at the value
+ *    type. The value type is looked up via `T[K]`. If it's not a (nested) object, the
+ *    value returns just the key `K`. So for a type `{a: string}`, this mapped type
+ *    results in `{a: 'a'}`
+ *
+ *    If the value type is an object though, then we recurse and we prefix the current
+ *    key. A type `{a: {b: string}}` then becomes (in steps):
+ *
+ *    1. `{a: 'a' | a.Paths<{b: string}>}`
+ *    2. `{a: 'a' | 'a.b'}` (because b only has a leaf node and returns `'b'``)
+ *
+ *  - 'finally', we grab all the values of the mapped type as a union via `[keyof T]`,
+ *    so this turns the mapped type `{a: 'a', b: 'b'}` into the union `'a' | 'b'`
+ *
+ *    When dealing with recursion/nested objects, the progression sort of looks like
+ *    the following for the example type `{a: {b: string, c: string}}`
+ *
+ *    1. `{a: 'a' | <prefix-and-recurse for key a>}`
+ *    2. recurse for a: `{b: 'b', c: 'c'}
+ *    3. grab the values (end of recursion reached): `{b: 'b', c: 'c'}['b' | 'c']` -> `'b' | 'c'`
+ *    4. exit recursion, apply the prefix on the recursion result: `'a.b' | 'a.c'`
+ *    5. join the access paths for a and from a: `{a: 'a' | ('a.b' | 'a.c')}`
+ *    6. grab the values via the indexed access: `{a: 'a' | ('a.b' | 'a.c')}['a']`
+ *    7. final result: `'a' | 'a.b' | 'a.c'`
+ *
+ *  - The `ExcludeBareObject` helper type is required because potentially the recursion
+ *    steps into a generic/bare `Object` type which recurses infintely, and TS can't deal
+ *    with that. Formio type definitions happen to use those in some places in their
+ *    source code and unfortunately those aren't easily patched/fixed.
+ */
+export type Paths<T> = T extends T
+  ? {
+      [K in keyof T]-?: K extends string
+        ? ExcludeBareObject<T[K]> extends object
+          ? K | `${K}.${Paths<ExcludeBareObject<T[K]>>}`
+          : K
+        : never;
+    }[keyof T]
+  : never;
```

Comment:
Seems to work if `T` is defined as `{[K: string]: ...}` (then `Paths<T>` is `string`)

However I've tried the following with `PathsForValueType`:

```ts
type Test = PathsForValueType<{[K: string]: Option[]}, Option[]>
// Test is string | number?
```

But haven't investigated further, still need to understand the other types below.

Reply:
I had to [play around](https://www.typescriptlang.org/play?ts=4.9.5#code/C4TwDgpgBAogHgYwDYFcAmEBCBDAThAeQCMArCBYAHgBUA+KAXimqgjmAgDs0BnKAJXIB7XGko9guAJacA5gBooKTgGtOQgO6daAKChQA-M1bsuvKADFlFKUM579hqJwgA3CLgf6AXMwe+Xd1wAbh0HUEgoAAVsYAALHhp6JhY2Dm4+agcjAG8vRwBtAGkoGSgVCBAhADNmAF0AWgNfErSzPglpOXzHJ3hkdCw8QlJyKmpiuvo2jKghUYoe3qcSgB8oAAMAEhyigF8AOh2Y+MT+1AwcfGIyChpJ2lo9jaXelteoALcPUN69goqVVq1Dq-mc3xCYX0EWgAHEIMAAGrYVAQACCwBOcRoiixJnS5k6MlkyWMM3MWX0Rjx5L42xyMmqHmihx2jOZggkz3y1Px7XKlRqfmWTnhSJRKHRmNi2ImUTqik5wF0Iq+QXyvhpplmgKFlMcRjlMtBb3B6p8Zp+YRh0RlPDR3GRqOo4AgSUYUDy+mKpU4ttO7uwfFa2sJkmJK0+ltwdV8YqdkoxWJxUCKtFCe1COhtWJ4FhECYgLsgKcRpN1tS9UAKUV9-oSgb48YlUuT1Fx01DfERTlrao8sagkklGazOhkHFw1WwCGgBDAwFsfqr+lcLd8RLkv0cSGwRAgSA34a3Oj21tdUAAykIALaEBdLm5jD0r6tFI9dWSD+eLuwFOqjuekQWB6uaUNed4-o+CzKqECB2BIUDVL4IFMAA5EQu6YdgAD02CcCA8TEmhWY2gQJRMLmDpoIWxZuhB96-pwT4UOmQHQNQECIZRdr5rgtGuuBt6MdBtzAIoUF-lMcEIcAQ5ccAvicdxUBoSR7FQAAipKHoVp6+Q1nWIFBlAzaokmMpCZBD52Cx4nRJ2BJ8JJnD-r2UaBAOvjDhAp46DhOFQDwQhQBoUjxEFx6yFAJkyBgcBDheMgSBA2BoHMtTYFASDhR4KJKJwS6KBocSxKFQgoEg6XxNA+A3tgYCQOlgLZUIQgqNlUgVAY2YXtpEAAEwemZibSvEVkibZMGKJuJJ8rMLluUYs0eRCoQBaUclIG1Kh8DlFSRZ+ACEvWRP1ADMumCpWxQfsSg4+ZmUAbSt6ycCgN77p4OEAFQ-XoP1QGiAASaJHUOcRSHwM7ACgKJICAUD1RUHRcDwECKPuCDYCg6NQFUKBQNjfoExD+DQDC0M8ADcwwb6aBSLOPB3XIijvZ9zL4elPAgJ9QhIAcV680Q-OlHwGgQFIogxdwRP4VA+6lLI6j4Gg8g0+FaHQ0gwVIdgUsI+CGgy+l6hyWAu6yJK6VEIjwNQ8AIiM-lmCXgcNM08QBTs19dQxRQcNIIb8GcEEwB8I7MWHRGE4eJw8MgIod74RG8MCiA1NQIDwzR3IPCC9eEPQDIQR4w1FtSFxQ5tVADShXEXD4xVUCle4MsxbgRDhbgeCI-MYnq1nTeE8TaFyWomhISIQUoPgENlTVvqTvHSC59FLVQ84QhyVlPseDNQg09U0+L3vuCJZE4VQGAuBCGgKBM2vUBvR9X0A4F-mBQAAuHDRsJAFA-64Fvp4EOiFj4hSYDkAAjL4aBA1zqKDQtAtCvg0J7gQGhTMOgwH8wgAcbasgAAUECACUWYNolTKlIWoEskZCDbovEarYZTVwEMDS8PUbRFAgB1JgelXyGTKBYQczCLLjQYi5OyHZ5rmEWn7Xk-ZPBnn8n9Gm1BcCI0jtUKQCUiJ5wVojba7UIwoDAEXKAa5UR1kXvwDhvoUppQyunKAdVy7Eg9kPA43it4xTXFIJAihF4bw6IuIORc55cwsXYy8YtozFUhggOIctOBjyChACAN5q400VjopAk4IDpQqnJbwRD34aUwLfConBcx8QEiWdsUAywvgcEIv0YE6CiIRIWcRsoZG0iae5JRgFxycEnNOWcV4DxjEvEkzJWVXwwjQejJAYx1K9DQLEbAl5cAIDQVYyUPB1mOE2cAbAvhXy9AOVxXw8jtz6DPCo2OU4ZzQH4GlWwsyG71X0r0JZqke4MyEMc1cLZmZQDuX5U60BLzTIoLmD0lT2pcFqQWFsdFwJwuAF8+ZEkbKuWkjg2S6TVnwplL4PSsLSVjQSB6NCpzsAHGuUcmSnBEIrLGFiAaFLrpTOpQi9CDKdmYOCPoDapSgpxAqlVFu2A277kbnkgpRSUDADCGAuSGgRAqGJFSzl5L6x5jRc6QSeqKA4vqnipi-5ST0q2UysF6kNoeBAVTEwACOBoCJWyzV2rdVYq5ZqXixrJQYrNdiuZlqIX4ptXSoVuySJAA) a bit to figure out why this produces that weird union type.

In short - when looking at a plain JS object with unspecified literal keys, `keyof Object` is `string | number` since both are valid to index the object. Number is converted to string, since internally only strings can be used to index.

`keyof {[K: string]: any} == string | number`

When order is respected in an object, try running this in the browser:

```js
console.log({123: 1, '123': 'a'});
// 'a'
```

Reply:
Ok makes sense now, another reason JS is such a broken language

Reply:
:thisisfine: